### PR TITLE
feat: bake gapic-generator-java into the hermetic build docker image

### PR DESCRIPTION
**Pull request by diegomarquezp**
Makes the gapic-generator-java jar a prepared binary in the Docker image whose location is now assumed by the scripts. Developers need now to prepare this well-know location (specified in `library_generation/DEVELOPMENT.md`.

### CODE REVIEW

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [inline .cloudbuild/library_generation/library_generation.Dockerfile:26]: Only copying the jar allows us for a more precise test. We can use the env vars `DOCKER_GAPIC_GENERATOR_LOCATION` and `DOCKER_GAPIC_GENERATOR_VERSION` to confirm if the preconfigured generator was used

File: .github/scripts/hermetic_library_generation.sh
Comment (diegomarquezp) [inline .github/scripts/hermetic_library_generation.sh:93]: there is a `docker build` call right after this deleted section. That image will contain the SNAPSHOT jar (and the corresponding env var)

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (blakeli0) [reply .cloudbuild/library_generation/library_generation.Dockerfile:26]: Does this jar have all the info we need? Can you test running the image without internet? Sometimes the dependencies of the jar are still downloaded.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (blakeli0) [inline .cloudbuild/library_generation/library_generation.Dockerfile:67]: Can we remove this version? Since we know the location from the previous stage, can we just use whatever jar is there?

File: library_generation/utils/utilities.sh
Comment (blakeli0) [inline library_generation/utils/utilities.sh:187]: I think this copy logic might already be covered [here](https://github.com/googleapis/sdk-platform-java/blob/6e20135d9777c221ee5c116b093a9c338ad24156/library_generation/utils/utilities.sh#L204-L205)?

File: library_generation/model/generation_config.py
Comment (blakeli0) [inline library_generation/model/generation_config.py:28]: Now that we have a `DOCKER_GAPIC_GENERATOR_VERSION` specified in the docker file, I think we can remove the usage of `GENERATOR_VERSION` now.

File: library_generation/utils/utilities.sh
Comment (blakeli0) [inline library_generation/utils/utilities.sh:161]: Can you please update the comment above as well?

File: library_generation/utils/utilities.sh
Comment (diegomarquezp) [reply library_generation/utils/utilities.sh:187]: I decided to not follow the `.m2` folder structure in favor of easier testing (https://github.com/googleapis/sdk-platform-java/pull/3067/files#r1698944411). I think it's ok to treat this baked-in jar as a special case outside of `.m2`?

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:26]: I ran `docker run --network none` and could get past `generate_library.sh` (i.e. no jars needed), but I found two things we may want to follow up with:

This function from the transferred synthtool fetches data from Maven:
https://github.com/googleapis/sdk-platform-java/blob/0ef66196744e21fc3b4e2553d817c0dea5efe127/library_generation/owlbot/synthtool/languages/java.py#L127

Our script pulls googleapis' proto definitions:
https://github.com/googleapis/sdk-platform-java/blob/0ef66196744e21fc3b4e2553d817c0dea5efe127/library_generation/utils/utilities.py#L175-L182


I thought there was a commit in synthtool that fetches libraries_bom from our config yaml?
https://github.com/googleapis/sdk-platform-java/blob/0ef66196744e21fc3b4e2553d817c0dea5efe127/library_generation/owlbot/synthtool/languages/java.py#L280-L287

For the googleapis download, I guess it's expected behavior

Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:26]: I did a temporary bypass (not committed) to see if the generation is successful. It is except for the `mvn fmt:format` part that needs to fetch data when processing a library. The data includes fetching the parent pom of the target library. Nothing about sdk-platform is fetched.

https://paste.googleplex.com/5092731556986880?raw

File: library_generation/model/generation_config.py
Comment (diegomarquezp) [reply library_generation/model/generation_config.py:28]: I thought I removed all of the references to the GENERATOR_VERSION env var but there was a missing one in `.github/scripts/hermetic_library_generation`. I also renamed the variable in line 26 to make it clear that it's a reference to the config yaml key

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:67]: https://github.com/googleapis/sdk-platform-java/blob/a5c18b2920bfd349119ed95a75da785d1fa3f23a/library_generation/utils/utilities.sh#L182-L187

File: library_generation/utils/utilities.sh
Comment (diegomarquezp) [reply library_generation/utils/utilities.sh:161]: Thank you, done.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (blakeli0) [reply .cloudbuild/library_generation/library_generation.Dockerfile:26]: Sounds good! Yes we should definitely but I don't think they should block this PR, since we already have all the jars in the docker image.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (blakeli0) [reply .cloudbuild/library_generation/library_generation.Dockerfile:67]: I see. For protoc, we tried to make it more flexible since we do have the need to test different protoc versions frequently. I don't think we have this need for gapic-generator-java. Hence I think it's better to make the implementation simple. It could be a separate PR also.

File: .github/scripts/hermetic_library_generation.sh
Comment (blakeli0) [inline .github/scripts/hermetic_library_generation.sh:110]: I think we can remove the `.m2` volume above also?

File: library_generation/model/generation_config.py
Comment (blakeli0) [reply library_generation/model/generation_config.py:28]: Thanks, this is much clearer now! As I mentioned in another comment, maybe we can try to get rid of `GAPIC_GENERATOR_VERSION_CONFIG_KEY` all together.

File: library_generation/utils/utilities.sh
Comment (blakeli0) [reply library_generation/utils/utilities.sh:192]: I see. Yes it's definitely OK to put it into a different location, as long as we don't use maven to retrieve it. Let's see if we can get rid of `gapic_generator_version` so that we don't need this whole block.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile]: ok, I removed gapic_generator_version from the scripts. We will now always assume `DOCKER_GAPIC_GENERATOR_LOCATION` has the jar we need. We don't care about the generator version anymore.

File: .github/scripts/hermetic_library_generation.sh
Comment (diegomarquezp) [reply .github/scripts/hermetic_library_generation.sh:110]: Done

File: library_generation/model/generation_config.py
Comment (diegomarquezp) [reply library_generation/model/generation_config.py:27]: Got rid of it

**Comment by diegomarquezp** [comment]
@blakeli0 I removed `gapic_generator_version` altogether. We now only rely on `DOCKER_GAPIC_GENERATOR_LOCATION` that will point to the jar we need and not care about the location.

This however brought many changes in the test folder but most of them are simple.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (blakeli0) [inline .cloudbuild/library_generation/library_generation.Dockerfile:68]: Do we need this location to be an env variable? Can this be inlined?

File: library_generation/utils/utilities.sh
Comment (blakeli0) [inline library_generation/utils/utilities.sh:178]: Can we just use the jar as it is without copying? Is it for `gapic-generator-java-wrapper`?

File: library_generation/test/resources/goldens/pr_description-golden.txt
Comment (blakeli0) [inline library_generation/test/resources/goldens/pr_description-golden.txt:5]: I think this is a feature we forgot, but we definitely still need it. There are three follow up changes that I can see

- Change [hermetic_library_generation.sh](https://github.com/googleapis/java-spanner/blob/main/.github/scripts/hermetic_library_generation.sh#L82) so that it does not parse from generation_config.yaml anymore.
- Change [update_generation_config.sh](https://github.com/googleapis/java-spanner/blob/main/.github/scripts/update_generation_config.sh#L97-L99) so that it automatically updates the docker image versions. 
- Change the commit generation logic so that it can detect the above version change and generate this commit message correctly. 

These changes can be in separate PRs but we need to make sure the existing automations are not breaking.

File: library_generation/utils/utilities.sh
Comment (diegomarquezp) [reply library_generation/utils/utilities.sh:178]: The wrapper assumes the jar to be in CWD

https://github.com/googleapis/sdk-platform-java/blob/58da069a99a7240cbec18b7dbdb010fb74c4d551/library_generation/gapic-generator-java-wrapper#L6

I'm not sure if we can both avoid copying the jar and at the same time not have a `DOCKER_GAPIC_GENERATOR_LOCATION` variable. We need a way to know where the jar is and make it compatible with local development as well.

Maybe we can copy the jar directly into the home `.m2` folder? We would still be copying it just like we've been doing so far.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:68]: We could set the jar into a well-known location. Developers would need to send their jar in there beforehand. 

I think this could solve the need for an env var and copying the jar?

In favor of the env var, I think it helps us separate the logic of the Dockerfile when it decides where to put the jar vs our scripts finding out where the jar is. But I'm not sure if this outweights the need to depend on an env var?

Copying the jar is more about keeping the changes minimal, without involving how the jar wrapper behaves. However, I think the reason we copy the jar into the output folder is because of the way we call the wrapper. 

https://github.com/googleapis/sdk-platform-java/blob/58da069a99a7240cbec18b7dbdb010fb74c4d551/library_generation/generate_library.sh#L209

We can configure the wrapper to expect the jar to be in the well-known location instead.

File: library_generation/test/resources/goldens/pr_description-golden.txt
Comment (blakeli0) [reply library_generation/test/resources/goldens/pr_description-golden.txt:5]: Discussed offline that we are going to keep using `gapic_generator_version` in generation_config.yaml. The usage will be limited to parsing the version for ocker image and generate commit history, not used for generate code.

File: library_generation/test/generate_library_unit_tests.sh
Comment (diegomarquezp) [inline library_generation/test/generate_library_unit_tests.sh:10]:
```suggestion
# we simulate a properly prepared environment (i.e. generator jar in its
# well-known location). Tests confirming the opposite case will make sure this
```

**Comment by diegomarquezp** [comment]
@JoeWang1127 @blakeli0 the integration tests now can't be locked to a specific generator version, so we will constantly need to update them. 

A couple ideas:
 * Add an `--update` flag to the integration tests
 * Keep a locked-in jar in the test resources folder

**Comment by diegomarquezp** [comment]
[spring downstream check failing](https://github.com/googleapis/sdk-platform-java/actions/runs/10625302690/job/29455529366) because the release job still doesn't publish the maven artifacts

https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/3138#issuecomment-2318598130

image


cc: @mpeddada1

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile]: We set the jar into a well-known location. No env vars used

File: library_generation/test/resources/goldens/pr_description-golden.txt
Comment (diegomarquezp) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: Done. Restored pr description generation.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (JoeWang1127) [inline .cloudbuild/library_generation/library_generation.Dockerfile:17]: Could we use image in Artifact Registry?

For example, https://airlock.corp.goog/package?type=Docker&id=088032b3-3669-43c0-9e82-02871abf2f0e

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (JoeWang1127) [reply .cloudbuild/library_generation/library_generation.Dockerfile:30]: Maybe as a follow-up.

File: library_generation/DEVELOPMENT.md
Comment (JoeWang1127) [inline library_generation/DEVELOPMENT.md:58]: Do you mean `${HOME}/.library_generation` needs to manually created before running the image?

File: library_generation/DEVELOPMENT.md
Comment (JoeWang1127) [inline library_generation/DEVELOPMENT.md:68]: When building an image, the generator jar is already put into the well-known location.

Why does a developer need to build and move the jar again?

File: library_generation/test/generate_library_unit_tests.sh
Comment (JoeWang1127) [inline library_generation/test/generate_library_unit_tests.sh:25]: Can we move the newly added shell test to `utilities_unit_tests.py` by using `util.sh_util()` method?

I think python unit tests are better maintained than shell tests. WDYT?

File: library_generation/utils/utilities.sh
Comment (JoeWang1127) [inline library_generation/utils/utilities.sh:143]: Could you remind me where to set `DOCKER_GAPIC_GENERATOR_LOCATION`?

I didn't find it in the development guide.

**Comment by JoeWang1127** [comment]
> the integration tests now can't be locked to a specific generator version, so we will constantly need to update them

I think it's fine to test the snapshot version of the generator. 

Ideally, running the generation can produce the same file even if the generator updates, though some changes in the generator can produce different code and we can update the golden file accordingly, just like the showcase integration test.

At this point, I think it's fine. However, we need to improvement the integration test:
- Use golden file rather than branch in google-cloud-java
- Provide an easy way to update the golden file

@blakeli0 WDYT?

File: library_generation/DEVELOPMENT.md
Comment (diegomarquezp) [reply library_generation/DEVELOPMENT.md:68]: They wouldn't need to prepare this jar when running the Docker image since the Dockerfile will take care of it. The thing is that this new section is meant for local development (it is under that section).

File: library_generation/DEVELOPMENT.md
Comment (diegomarquezp) [reply library_generation/DEVELOPMENT.md:58]: No, this guide is part of the local development instructions. The Dockerfile will take care of preparing this aspect of the environment.

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:17]: I tried to scope this to baking the generator only. The next step in the milestone is to "cleanup" the Docker image, involving several improvements including using safer base images. Maybe we can leave it to the next task, WDYT?

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:30]: Yes, agreed it can be a follow up. I explained this is planned in our current milestone (https://github.com/googleapis/sdk-platform-java/pull/3067#discussion_r1739185627).

File: library_generation/test/generate_library_unit_tests.sh
Comment (diegomarquezp) [reply library_generation/test/generate_library_unit_tests.sh:25]: I was under the impression that `genreate_library_units` is meant for `utilities.sh` and that `utilities_units.py` was for `utilities.py`. 

Could you please explain how are they `py` unit tests better maintained?

File: library_generation/utils/utilities.sh
Comment (diegomarquezp) [reply library_generation/utils/utilities.sh:143]: Ah, this is wrong. I updated the comment to explain about the well-known location. Thanks for the catch!

**Comment by diegomarquezp** [comment]
From discussion with @blakeli0 and @JoeWang1127, we can lock in a specific version string of the generator, download it in our integration test and then override the well-known location via volumes

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (JoeWang1127) [reply .cloudbuild/library_generation/library_generation.Dockerfile:17]: SGTM

**Comment by diegomarquezp** [comment]
@JoeWang1127 I modified the integration test to lock-in a specific generator jar downloaded via `mvn`. PTAL when you have a chance :)

File: .kokoro/presubmit/downstream-compatibility-spring.sh
Comment (diegomarquezp) [inline .kokoro/presubmit/downstream-compatibility-spring.sh:54]: @blakeli0 @JoeWang1127  I fixed the spring check in this PR as well since I thought it's small enough. There is a TODO to be solved next week, once https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/3183 is merged

File: library_generation/DEVELOPMENT.md
Comment (JoeWang1127) [reply library_generation/DEVELOPMENT.md:58]: Ah, got it.

**Comment by JoeWang1127** [comment]
> @JoeWang1127 I modified the integration test to lock-in a specific generator jar downloaded via `mvn`. PTAL when you have a chance :)

Thanks, could you change the pr body since this is no longer a pending issue.

File: library_generation/test/generate_library_unit_tests.sh
Comment (JoeWang1127) [reply library_generation/test/generate_library_unit_tests.sh:25]: > Could you please explain how are the py unit tests better maintained?

The shell script doesn't have (or we don't use) a testing framework so we need to implement our own testing utilities, e.g., assert, test counters, etc, in `test_utilities.sh`. Every time we want to add a test, we need to add the test function to the test list.

All these burden was introduced at the beginning stage of this project because we didn't use python at that time.

Right now we have a python language framework already setup and we should migrate all shell script (with tests) to python, although this is out of scope of this PR.

> I was under the impression that genreate_library_unit_tests.sh was meant for utilities.sh and that utilities_units.py was for utilities.py.

For newly added shell tests, I suggest to wrap the test with `util.sh_util()`.

I didn't notice the test is verifying `genreate_library.sh`. Create `genreate_library_unit_tests.py` and put the test there is better. WDYT?

File: library_generation/DEVELOPMENT.md
Comment (JoeWang1127) [inline library_generation/DEVELOPMENT.md:58]: nit: one sentence per line.

File: library_generation/test/integration_tests.py
Comment (JoeWang1127) [inline library_generation/test/integration_tests.py:56]:
```suggestion
# See library_generation/DEVELOPMENT.md
```

Comment (JoeWang1127) [reply library_generation/test/integration_tests.py:56]: Could you specify which part of `development.md` you are refering?

Comment (diegomarquezp) [reply library_generation/test/integration_tests.py:56]: Done.

File: library_generation/DEVELOPMENT.md
Comment (diegomarquezp) [reply library_generation/DEVELOPMENT.md]: I found [this article](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line) that confirms this is a good practice. I reformatted the file.

File: .kokoro/presubmit/downstream-compatibility-spring.sh
Comment (diegomarquezp) [inline .kokoro/presubmit/downstream-compatibility-spring.sh:53]:
```suggestion
# https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/3183 is merged and released.
```

File: .cloudbuild/library_generation/library_generation.Dockerfile
Comment (JoeWang1127) [inline .cloudbuild/library_generation/library_generation.Dockerfile:70]: A future improvement: only include necessary files in the final image.

Comment (diegomarquezp) [reply .cloudbuild/library_generation/library_generation.Dockerfile:70]: Yes, we want to get rid of the source files (owlbot cli, sdk-platform, etc) by building in this stage then only copying the binaries. This is part of the image cleanup task.

File: library_generation/DEVELOPMENT.md
Comment (blakeli0) [inline library_generation/DEVELOPMENT.md:57]: I would prefer not mention this as this is not implemented yet. We may treat gRPC and protoc differently as well.

Comment (diegomarquezp) [reply library_generation/DEVELOPMENT.md]: Got it, I removed this sentence.

File: library_generation/test/resources/integration/.gitignore
Comment (blakeli0) [inline library_generation/test/resources/integration/.gitignore:1]: Can we move this to the `.gitignore` under the [root folder](https://github.com/googleapis/sdk-platform-java/blob/main/.gitignore#L22-L26)?

File: library_generation/test/integration_tests.py
Comment (blakeli0) [inline library_generation/test/integration_tests.py:298]: I guess this would override the SNAPSHOT jar built in the previous step `docker build`?

Comment (diegomarquezp) [reply library_generation/test/integration_tests.py:298]: Yes, exactly. This is just to ensure that the outcome of the test doesn't depend on changes to the generator code.

File: library_generation/test/resources/integration/.gitignore
Comment (diegomarquezp) [reply library_generation/test/resources/integration/.gitignore:1]: Done

**Review by blakeli0 — APPROVED** [review]

**Review by JoeWang1127 — APPROVED** [review]

**Pull request merged**
